File: app/channels/chat_rooms_channel.rb
# frozen_string_literal: true

# ChatRooms Channel
class ChatRoomsChannel < ApplicationCable::Channel
  def subscribed
    stream_from "chat_rooms_channel_#{params[:chat_room_id]}" unless params[:chat_room_id].nil?
  end

  def unsubscribed; end

  def seen(data)
    chat_room = ChatRoom.find(data['chat_room_id'])
    messages = chat_room.messages.where.not(user: current_user).where(seen_at: nil)
    messages.map(&:seen!)
    other_user = chat_room.user1 == current_user ? chat_room.user2 : chat_room.user1
    notifications = Notification.all.where(recipient: current_user, actor: other_user, seen_at: nil)
    notifications.map(&:seen!)
    ActionCable.server.broadcast "chat_rooms_channel_#{data['chat_room_id']}",
                                 type: 'seen', messages_id: messages.pluck(:id)
  end
end

File: app/javascript/channels/chat_rooms_channel.js
import consumer from "./consumer"

$(document).on("turbolinks:load", function () {
  const chat_room_id = $('#message_chat_room_id').val()
  const other_user = $('#chat-username').attr('data-username')

  if (this.chat_subscription) {
    consumer.subscriptions.remove(this.chat_subscription);
    this.chat_subscription = null
  }

  function updateScroll() {
    const element = document.getElementById('messages');
    element.scrollTop = element.scrollHeight;
  }

  if (chat_room_id) {
    let connection = consumer.subscriptions.create({ channel: "ChatRoomsChannel", chat_room_id: chat_room_id }, {
      connected() {
        this.perform('seen', { chat_room_id: chat_room_id })
      },

      disconnected() {
        // Called when the subscription has been terminated by the server
      },

      received(data) {
        if (data.type == 'creation') {
          if (data.user == other_user) {
            if ($('#messages').length) {
              $('#messages').append(`${data.html_other_user}`)
              this.perform('seen', { chat_room_id: chat_room_id })
            }
          } else {
            $('#messages').append(`${data.html_current_user}`)
          }
          updateScroll()
          $('#message_text').val('')
        } else if (data.type == 'seen') {
          data.messages_id.forEach(id => {
            $(`#message-${id}-seen`).addClass('text-success')
          })
        } else {
          const message_id = `#message-${data.message_id}`
          $(`${message_id} p:first-child`).html(
            '<i class="fas fa-ban"></i> This message was deleted'
          )
          $(`${message_id} a`).remove()
          $(`${message_id}-seen`).remove()
        }
      }
    })

    this.chat_subscription = connection
  }
})
